### ISSUE TITLE
Syncing from the sequencer, the RPC node is behind because the datastream lags (both blocks and batches)

### ISSUE THREAD

**Issue by mandrigin** [0xPolygon/cdk-erigon#813]
New blocks written to the datastream are not immediately shown in the node's datastream (even when queried with Toni Ramirez's tool). When we write block N to the datastream, it is only updated with the block N-1, sometimes N-2 at the tip.

UPD: same is true for the batch-by-batch approach, see https://github.com/0xPolygonHermez/cdk-erigon/issues/813#issuecomment-2238900102

**Comment by hexoscott** [comment]
To add in a little detail on this, when we get to the point of pushing a single block at a time to the stream we:
- open an atomic operation
- write a batch start bookmark and batch start entry
- write the block bookmark
- write the block header info
- write transactions
- if it is the end of a batch we write the batch end
- commit the atomic operation

So we appear to be doing everything correctly but the stream consumers, both RPC and the datastream tool, do not pick up on this information until something else is written to the stream.  It looks like it isn't "published" immediately.

I tested a really long block time of 15 seconds to trial if it is just a timing problem but the same issue, the RPC lagged one block behind as did the datastream tool.

We are using v0.2.2 of the datastreamer library in cdk-erigon.

**Comment by mandrigin** [comment]
some extra data, is that the delay (in seconds) increases if we increase block time, and delay (in blocks) stays around constant. might be important.

**Comment by mandrigin** [comment]
you can try to reproduce
(1) zkevm branch **without** stateless executor and executor-strict: false
(2) https://github.com/0xPolygonHermez/cdk-erigon/pull/797 with any scenario (with or without stateless executor)

**Comment by mandrigin** [comment]
Same happens on batch-by-batch.

See DataDog from Bali, and see quite a delay in DS update, until it fixes itself eventually.

<img width="1574" alt="image" src="https://github.com/user-attachments/assets/00e94fd7-32bb-42f5-a247-c9767826c980">

**Comment by V-Staykov** [comment]
One problem I found and now fixed is that upon reading a block proto, we enter a loop to read after it more files. This is left from previous datastream versions, where txs were written on separate files and thus we needed to keep reading until we get a block end, so we have all the transactions in the block as well. This lead to the current situation where we read blocks until we get next block's bookmark at least. basically we were waiting for the next block to be produced.

Currently all this is not needed and we can process datastream 1 file at a time.
